network/test/middleware_test.go

```diff
@@ -174,6 +182,200 @@ func (m *MiddlewareTestSuite) TestUpdateNodeAddresses() {
 	require.NoError(m.T(), err)
 }
 
+func (m *MiddlewareTestSuite) TestUnicastRateLimit_Messages() {
+	// limiter limit will be set to 5 events/sec the 6th event per interval will be rate limited
+	limit := rate.Limit(5)
+
+	// burst per interval
+	burst := 5
+
+	// create test time
+	testtime := unittest.NewTestTime()
+
+	// setup message rate limiter
+	messageRateLimiter := unicast.NewMessageRateLimiter(limit, burst, 1, unicast.WithGetTimeNowFunc(testtime.Now))
+
+	// the onUnicastRateLimitedPeerFunc call back we will use to keep track of how many times a rate limit happens
+	// after 5 rate limits we will close ch. O
+	ch := make(chan struct{})
+	var rateLimits uint64
+	onRateLimit := func(peerID peer.ID, role, msgType string, topic channels.Topic, reason unicast.RateLimitReason) {
+		require.Equal(m.T(), reason, unicast.MessageCount)
+
+		// we only expect messages from the first middleware on the test suite
+		expectedPID, err := unittest.PeerIDFromFlowID(m.ids[0])
+		require.NoError(m.T(), err)
+		require.Equal(m.T(), expectedPID, peerID)
+
+		// update hook calls
+		atomic.AddUint64(&rateLimits, 1)
+	}
+
+	rateLimiters := unicast.NewRateLimiters(messageRateLimiter, &unicast.NoopRateLimiter{}, onRateLimit, false)
+
+	// create a new staked identity
+	ids, libP2PNodes, _ := GenerateIDs(m.T(), m.logger, 1)
+	defer testnet.StopNodes(m.T(), libP2PNodes)
+
+	// create middleware
+	netmet := mock.NewNetworkMetrics(m.T())
+	calls := 0
+	netmet.On("NetworkMessageReceived", mockery.Anything, mockery.Anything, mockery.Anything).Times(5).Run(func(args mockery.Arguments) {
+		calls++
+		if calls == 5 {
+			close(ch)
+		}
+	})
+	// we expect 5 messages to be processed the rest will be rate limited
+	defer netmet.AssertNumberOfCalls(m.T(), "NetworkMessageReceived", 5)
+
+	mws, providers := GenerateMiddlewares(m.T(), m.logger, ids, libP2PNodes, unittest.NetworkCodec(), m.slashingViolationsConsumer, WithUnicastRateLimiters(rateLimiters), WithNetworkMetrics(netmet))
+	require.Len(m.T(), ids, 1)
+	require.Len(m.T(), providers, 1)
+	require.Len(m.T(), mws, 1)
+	newId := ids[0]
+	newMw := mws[0]
+
+	overlay := m.createOverlay(providers[0])
+	overlay.On("Receive",
+		m.ids[0].NodeID,
+		mockery.AnythingOfType("*message.Message"),
+	).Return(nil)
+
+	newMw.SetOverlay(overlay)
+
+	ctx, cancel := context.WithCancel(m.mwCtx)
+	irrecoverableCtx, _ := irrecoverable.WithSignaler(ctx)
+	newMw.Start(irrecoverableCtx)
+
+	idList := flow.IdentityList(append(m.ids, newId))
+
+	// needed to enable ID translation
+	m.providers[0].SetIdentities(idList)
+
+	// update the addresses
+	m.mws[0].UpdateNodeAddresses()
+
+	// for the duration of a simulated second we will send 10 messages. The 10th message will be
+	// rate limited.
+	start := testtime.Now()
+	end := start.Add(time.Second)
+	for testtime.Now().Before(end) {
+		// a message is sent every 167 milliseconds which equates to around 6 req/sec surpassing our limit
+		testtime.Advance(168 * time.Millisecond)
```

Comment:
this is going to make the test run slower than necessary. Could we just send 6 messages as fast as possible, and make sure the 6th is rate limited?

Reply:
The reason we do this is so that we can control the time , we need explicit control or else the test will run to fast. At most this will take 1 second.